can't close webpack-dev-server --progress with Ctrl + C

<!--
  Please note that this template is not optional.
  Please fill out _ALL_ fields, or your issue may be closed as "invalid."
  Please do not delete this template.
  Please ask questions on StackOverflow or Gitter (https://gitter.im/webpack/webpack).
  General questions, how-to questions, and support requests will be closed.
-->

* Operating System: macOS High Sierra (10.13.6)
* Node Version: v10.9.0
* NPM Version: 6.2.0
* webpack Version: 4.16.5
* webpack-dev-server Version: 3.1.5

<!--
  Please place an x, no spaces, in all [ ] that apply
  Please note that we are NOT accepting FEATURE requests at this time.
-->

- [x] This is a **bug**
- [ ] This is a **modification** request

### Code

Running the dev server with: `yarn webpack-dev-server --progress`. I'm not including webpack configs since seems to be reproducing on a variety of projects, so I assume it's something that happens regardless of the config? (Happy to provide more info otherwise).

### Expected Behavior

Ctrl + C stops the dev server.

### Actual Behavior

Ctrl + C does not do anything, dev server continues to compile modules.

_May be a regression of #860, #1360?_

have you guys found this error when press ctrl+C
```bash
if (serverData.server) {
                     ^
TypeError: Cannot read property 'server' of undefined
    at process.on (/Users/liangjiawen/workspace/recore/test/node_modules/webpack-dev-server/lib/utils/setupExitSignals.js:8:22)
    at process.emit (events.js:194:15)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
```
webpack-dev-server@3.9.0

--- a/lib/Server.js
+++ b/lib/Server.js
@@ -1114,11 +1114,26 @@ class Server {
     if (this.options.setupExitSignals) {
       const signals = ["SIGINT", "SIGTERM"];
 
+      let needForceShutdown = false;
+
       signals.forEach((signal) => {
         process.on(signal, () => {
-          this.stopCallback(() => {
+          if (needForceShutdown) {
             // eslint-disable-next-line no-process-exit
             process.exit();
+          }
+
+          this.logger.info(
+            "Gracefully shutting down. To force exit, press ^C again. Please wait..."
+          );
+
+          needForceShutdown = true;
+
+          this.stopCallback(() => {
+            this.compiler.close(() => {
+              // eslint-disable-next-line no-process-exit
+              process.exit();
+            });
           });
         });
       });

--- a/lib/Server.js
+++ b/lib/Server.js
@@ -1130,10 +1130,15 @@ class Server {
           needForceShutdown = true;
 
           this.stopCallback(() => {
-            this.compiler.close(() => {
+            if (this.compiler.close) {
+              this.compiler.close(() => {
+                // eslint-disable-next-line no-process-exit
+                process.exit();
+              });
+            } else {
               // eslint-disable-next-line no-process-exit
               process.exit();
-            });
+            }
           });
         });
       });

--- a/lib/Server.js
+++ b/lib/Server.js
@@ -1116,11 +1116,15 @@ class Server {
 
       let needForceShutdown = false;
 
+      const exitProcess = () => {
+        // eslint-disable-next-line no-process-exit
+        process.exit();
+      };
+
       signals.forEach((signal) => {
         process.on(signal, () => {
           if (needForceShutdown) {
-            // eslint-disable-next-line no-process-exit
-            process.exit();
+            exitProcess();
           }
 
           this.logger.info(
@@ -1130,14 +1134,10 @@ class Server {
           needForceShutdown = true;
 
           this.stopCallback(() => {
-            if (this.compiler.close) {
-              this.compiler.close(() => {
-                // eslint-disable-next-line no-process-exit
-                process.exit();
-              });
+            if (typeof this.compiler.close === "function") {
+              this.compiler.close(exitProcess);
             } else {
-              // eslint-disable-next-line no-process-exit
-              process.exit();
+              exitProcess();
             }
           });
         });

--- a/test/server/setupExitSignals-option.test.js
+++ b/test/server/setupExitSignals-option.test.js
@@ -37,7 +37,10 @@ describe("setupExitSignals option", () => {
       .spyOn(process.stdin, "resume")
       .mockImplementation(() => {});
     stopCallbackSpy = jest.spyOn(server, "stopCallback");
-    closeCallbackSpy = jest.spyOn(server.compiler, "close");
+
+    if (server.compiler.close) {
+      closeCallbackSpy = jest.spyOn(server.compiler, "close");
+    }
   });
 
   afterEach(async () => {
@@ -58,7 +61,10 @@ describe("setupExitSignals option", () => {
       const interval = setInterval(() => {
         if (doExit) {
           expect(stopCallbackSpy.mock.calls.length).toEqual(1);
-          expect(closeCallbackSpy.mock.calls.length).toEqual(1);
+
+          if (server.compiler.close) {
+            expect(closeCallbackSpy.mock.calls.length).toEqual(1);
+          }
 
           clearInterval(interval);
 

--- a/test/server/setupExitSignals-option.test.js
+++ b/test/server/setupExitSignals-option.test.js
@@ -11,6 +11,7 @@ describe("setupExitSignals option", () => {
   let exitSpy;
   let stopCallbackSpy;
   let stdinResumeSpy;
+  let closeCallbackSpy;
 
   const signals = ["SIGINT", "SIGTERM"];
 
@@ -36,6 +37,7 @@ describe("setupExitSignals option", () => {
       .spyOn(process.stdin, "resume")
       .mockImplementation(() => {});
     stopCallbackSpy = jest.spyOn(server, "stopCallback");
+    closeCallbackSpy = jest.spyOn(server.compiler, "close");
   });
 
   afterEach(async () => {
@@ -56,6 +58,7 @@ describe("setupExitSignals option", () => {
       const interval = setInterval(() => {
         if (doExit) {
           expect(stopCallbackSpy.mock.calls.length).toEqual(1);
+          expect(closeCallbackSpy.mock.calls.length).toEqual(1);
 
           clearInterval(interval);
 

